tensorflow_quantum/core/ops/noise/tfq_noisy_expectation.cc

```diff
@@ -0,0 +1,391 @@
+      auto sv = ss.Create(largest_nq);
+      auto scratch = ss.Create(largest_nq);
+
+      for (int i = 0; i < ncircuits.size(); i++) {
+        int nq = num_qubits[i];
+
+        // (#679) Just ignore empty program
+        if (ncircuits[i].channels.size() == 0) {
+          for (int j = 0; j < pauli_sums[i].size(); j++) {
+            (*output_tensor)(i, j) = -2.0;
+          }
+          continue;
+        }
+
+        if (nq > largest_nq) {
+          largest_nq = nq;
+          sv = ss.Create(largest_nq);
+          scratch = ss.Create(largest_nq);
+        }
+        QTSimulator::Parameter param;
+        param.collect_kop_stat = false;
+        param.collect_mea_stat = false;
+        param.normalize_before_mea_gates = true;
+        std::vector<uint64_t> unused_stats;
+        // Track op-wise stats.
+        std::vector<int> run_samples(num_samples[i].size(), 0);
+        std::vector<double> rolling_sums(num_samples[i].size(), 0.0);
+
+        while (1) {
+          ss.SetStateZero(sv);
+          // time since epoch seeds random generator.
+          unsigned long r_seed =
+              std::chrono::duration_cast<std::chrono::milliseconds>(
+                  std::chrono::system_clock::now().time_since_epoch())
+                  .count();
+
+          QTSimulator::RunOnce(param, ncircuits[i], r_seed, ss, sim, scratch,
+                               sv, unused_stats);
+
+          // Compute expectations across all ops using this trajectory.
+          for (int j = 0; j < pauli_sums[i].size(); j++) {
+            if (run_samples[j] >=
+                std::max(num_samples[i][j] / num_threads, 1)) {
+              continue;
+            }
+            float exp_v = 0.0;
+            OP_REQUIRES_OK(context,
+                           ComputeExpectationQsim(pauli_sums[i][j], sim, ss, sv,
+                                                  scratch, &exp_v));
+            rolling_sums[j] += static_cast<double>(exp_v);
+            run_samples[j]++;
+          }
+
+          // Check if we have run enough trajectories for all ops.
+          bool break_loop = true;
+          for (int j = 0; j < num_samples[i].size(); j++) {
+            if (run_samples[j] < std::max(num_samples[i][j] / num_threads, 1)) {
+              break_loop = false;
+              break;
+            }
+          }
+          if (break_loop) {
+            // Lock writing to this batch index in output_tensor.
+            batch_locks[i].lock();
+            for (int j = 0; j < num_samples[i].size(); j++) {
+              int divisor = num_samples[i][j];
+              // account for when num_samples[i][j] < num_threads.
+              if (num_samples[i][j] / num_threads == 0) {
+                divisor = num_threads;
+              }
+              rolling_sums[j] /= divisor;
+              (*output_tensor)(i, j) += static_cast<float>(rolling_sums[j]);
+            }
+            batch_locks[i].unlock();
+            break;
```

Comment:
This averaging seems to assume that if `num_samples[i][j] < num_threads`, run expectation once (by `run_samples[j] >= std::max(num_samples[i][j] / num_threads, 1)`) for all `num_threads` threads regardless of what `num_samples[i][j]` is, and then average over the `num_threads` results. Is it intended?

Reply:
Also, regardless of the above question, it seems work with the multiple of `num_threads`. For example, if we say `num_samples[i][j] = 2 * num_threads + 50` and `num_threads=100`. Since `num_samples[i][j] / num_threads == 0` is false, we will divide the `rolling_sums[j]` by `num_samples[i][j]`, which is `250`. However, the `100` threads will run the `std::max(num_samples[i][j] / num_threads, 1) = std::max(250 / 100, 1) = 2`, so the remainder `50` runs are ignored, which can be a reason of discrepancy. Is it intended?

Reply:
Oops, good catch. I'm going to try and think of a better way to schedule the work more exactly. It will be tricky to schedule the work so that each thread gets roughly the same amount while now accounting for some threads maybe doing `k` reps of the circuit and others doing `k + 1` reps.

Reply:
Thanks, the new comments on `num_samples` make it clear!